src/linear_algebra/invariant_submodule.lean

```diff
@@ -0,0 +1,125 @@
+/-
+Copyright (c) 2023 Monica Omar. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Monica Omar
+-/
+import linear_algebra.basic
+import linear_algebra.projection
+
+/-!
+# Invariant submodules
+
+In this file, we define and prove some basic results on invariant submodules.
+-/
+
+namespace submodule
+
+variables {E R : Type*} [ring R] [add_comm_group E] [module R E]
+
+/-- `U` is `T` invariant : `U ≤ U.comap` -/
+def invariant_under (U : submodule R E) (T : E →ₗ[R] E) : Prop := U ≤ U.comap T
+
+/-- `U` is `T` invariant if and only if `U.map T ≤ U` -/
+lemma invariant_under_iff_map (U : submodule R E) (T : E →ₗ[R] E) :
+  U.invariant_under T ↔ U.map T ≤ U := submodule.map_le_iff_le_comap.symm
+
+/-- `U` is `T` invariant if and only if `set.maps_to T U U` -/
+lemma invariant_under_iff_maps_to (U : submodule R E) (T : E →ₗ[R] E) :
+  U.invariant_under T ↔ set.maps_to T U U := iff.rfl
+
+/-- `U` is `T` invariant is equivalent to saying `T(U) ⊆ U` -/
+lemma invariant_under_iff (U : submodule R E) (T : E →ₗ[R] E) :
+  U.invariant_under T ↔ T '' U ⊆ U := by rw [← set.maps_to', U.invariant_under_iff_maps_to]
+
+variables (U V : submodule R E) (hUV : is_compl U V) (T : E →ₗ[R] E)
+
+local notation `pᵤ` := submodule.linear_proj_of_is_compl U V hUV
+local notation `pᵥ` := submodule.linear_proj_of_is_compl V U hUV.symm
+
+lemma proj_comp_self_comp_proj_eq_of_invariant_under
+  (h : U.invariant_under T) (x : E) : ↑(pᵤ (T ↑(pᵤ x))) = T ↑(pᵤ x) :=
+begin
+  rw submodule.linear_proj_of_is_compl_eq_self_iff,
+  exact h (submodule.coe_mem _),
+end
+
+lemma invariant_under_of_proj_comp_self_comp_proj_eq
+  (h : ∀ x : E, ↑(pᵤ (T ↑(pᵤ x))) = T ↑(pᵤ x)) : U.invariant_under T :=
+begin
+  intros u hu,
+  rw [submodule.mem_comap, ← submodule.linear_proj_of_is_compl_eq_self_iff hUV _,
+      ← (submodule.linear_proj_of_is_compl_eq_self_iff hUV u).mpr hu, h],
```

Comment:
you can remove `submodule.` here and elsewhere since you're already in the `submodule` namespace.
```suggestion
  rw [mem_comap, ← linear_proj_of_is_compl_eq_self_iff hUV,
      ← (linear_proj_of_is_compl_eq_self_iff hUV u).mpr hu, h],
```